File: GameAssistant/web/frontend/src/api/training.ts
import request from './request'

export interface TrainingJob {
  id: string
  name: string
  dataset_version_id: string
  base_model_architecture: string
  status: string
  epochs: number
  batch_size: number
  img_size: number
  lr0: number
  lrf: number
  weight_decay: number
  momentum: number
  patience: number
  mosaic: number
  mixup: number
  hsv_h: number
  hsv_s: number
  hsv_v: number
  flip_lr: number
  resume_from?: string
  best_model_id?: string
  process_id?: number
  log_output?: string
  log_summary?: Record<string, unknown>
  current_epoch?: number
  gpu_device?: string
  started_at?: string
  completed_at?: string
  created_by?: string
  created_at: string
  error_message?: string
}

export interface TrainingLog {
  id: string
  training_job_id: string
  epoch: number
  train_box_loss?: number
  train_cls_loss?: number
  train_dfl_loss?: number
  val_box_loss?: number
  val_cls_loss?: number
  val_dfl_loss?: number
  precision?: number
  recall?: number
  map50?: number
  map50_95?: number
  lr?: number
  gpu_memory_mb?: number
  epoch_duration_sec?: number
  logged_at: string
}

export interface TrainingProgress {
  job_id: string
  status: string
  current_epoch: number
  total_epochs: number
  best_model_id?: string
  current_map50?: number
  current_map?: number
  latest_metrics?: TrainingLog
}

export const trainingApi = {
  async getList(params?: {
    page?: number
    page_size?: number
    status?: string
    dataset_version_id?: string
  }): Promise<TrainingJob[]> {
    const q = new URLSearchParams()
    if (params?.page) q.set('page', String(params.page))
    if (params?.page_size) q.set('page_size', String(params.page_size))
    if (params?.status) q.set('status', params.status)
    if (params?.dataset_version_id) q.set('dataset_version_id', params.dataset_version_id)
    return request.get(`/training/${q.toString() ? `?${q}` : ''}`)
  },

  async create(data: Partial<TrainingJob>): Promise<TrainingJob> {
    return request.post('/training/', data)
  },

  async getById(id: string): Promise<TrainingJob> {
    return request.get(`/training/${id}`)
  },

  async update(id: string, data: Partial<TrainingJob>): Promise<TrainingJob> {
    return request.put(`/training/${id}`, data)
  },

  async stop(id: string): Promise<{ success: boolean }> {
    return request.post(`/training/${id}/stop`)
  },

  async getMetrics(jobId: string): Promise<TrainingLog[]> {
    return request.get(`/training/${jobId}/metrics`)
  },

  async getProgress(jobId: string): Promise<TrainingProgress> {
    return request.get(`/training/${jobId}/progress`)
  },

  getLogsSSE(jobId: string): string {
    return `/api/v1/training/${jobId}/logs`
  },
}

export default trainingApi

File: GameAssistant/web/backend/app/schemas/training.py
"""Pydantic schemas for training management."""
from datetime import datetime
from typing import Optional
from pydantic import Field, ConfigDict

from app.schemas._base import NoProtectedNamespaceModel


class TrainingJobBase(NoProtectedNamespaceModel):
    name: str
    dataset_version_id: str
    base_model_architecture: str = "yolov8n"
    epochs: int = 50
    batch_size: int = 8
    img_size: int = 640
    lr0: float = 0.01
    lrf: float = 0.01
    weight_decay: float = 0.0005
    momentum: float = 0.937
    patience: int = 15
    mosaic: float = 1.0
    mixup: float = 0.0
    hsv_h: float = 0.015
    hsv_s: float = 0.7
    hsv_v: float = 0.4
    flip_lr: float = 0.5


class TrainingJobCreate(TrainingJobBase):
    resume_from: Optional[str] = None


class TrainingJobUpdate(NoProtectedNamespaceModel):
    name: Optional[str] = None
    status: Optional[str] = None
    current_epoch: Optional[int] = None
    best_model_id: Optional[str] = None
    error_message: Optional[str] = None


class TrainingJobResponse(TrainingJobBase):
    id: str
    status: str
    resume_from: Optional[str] = None
    best_model_id: Optional[str] = None
    process_id: Optional[int] = None
    log_output: Optional[str] = None
    log_summary: Optional[dict] = None
    current_epoch: Optional[int] = None
    gpu_device: Optional[str] = None
    started_at: Optional[datetime] = None
    completed_at: Optional[datetime] = None
    created_by: Optional[str] = None
    created_at: datetime
    error_message: Optional[str] = None

    model_config = ConfigDict(from_attributes=True, protected_namespaces=())


class TrainingLogResponse(NoProtectedNamespaceModel):
    id: str
    training_job_id: str
    epoch: int
    train_box_loss: Optional[float] = None
    train_cls_loss: Optional[float] = None
    train_dfl_loss: Optional[float] = None
    val_box_loss: Optional[float] = None
    val_cls_loss: Optional[float] = None
    val_dfl_loss: Optional[float] = None
    precision: Optional[float] = None
    recall: Optional[float] = None
    map50: Optional[float] = None
    map50_95: Optional[float] = None
    lr: Optional[float] = None
    gpu_memory_mb: Optional[int] = None
    epoch_duration_sec: Optional[int] = None
    logged_at: datetime

    model_config = ConfigDict(from_attributes=True, protected_namespaces=())


class TrainingProgressResponse(NoProtectedNamespaceModel):
    job_id: str
    status: str
    current_epoch: int
    total_epochs: int
    best_model_id: Optional[str] = None
    current_map50: Optional[float] = None
    current_map: Optional[float] = None
    latest_metrics: Optional[TrainingLogResponse] = None
